v2/selective.go

```diff
@@ -0,0 +1,268 @@
+		ctx:      ctx,
+		root:     root,
+		selector: selector,
+		ls:       ls,
+		opts:     opts,
+	}
+	return &tc, nil
+}
+
+// FileTraversal writes a car file matching a given root and selector to the
+// path at `destination` using one read of each block.
+func FileTraversal(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, destination string, opts ...Option) error {
+	tc := traversalCar{
+		size:     0,
+		ctx:      ctx,
+		root:     root,
+		selector: selector,
+		ls:       ls,
+		opts:     opts,
+	}
+
+	fp, err := os.Create(destination)
+	if err != nil {
+		return err
+	}
+	defer fp.Close()
+
+	_, err = tc.WriteTo(fp)
+	if err != nil {
+		return err
+	}
+
+	// fix header size.
+	if _, err = fp.Seek(0, 0); err != nil {
+		return err
+	}
+
+	tc.size = uint64(tc.size)
+	if _, err = tc.WriteV2Header(fp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// V1Traversal walks through the proposed dag traversal and writes a carv1 to the provided io.Writer
+func V1Traversal(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, writer io.Writer, opts ...Option) (uint64, error) {
+	opts = append(opts, WithoutIndex())
+	tc := traversalCar{
+		size:     0,
+		ctx:      ctx,
+		root:     root,
+		selector: selector,
+		ls:       ls,
+		opts:     opts,
+	}
+
+	len, _, err := tc.WriteV1(writer)
+	return len, err
+}
+
+// Writer is an interface allowing writing a car with the data specified by a selector.
+type Writer interface {
+	io.WriterTo
+}
+
+type traversalCar struct {
+	size     uint64
+	ctx      context.Context
+	root     cid.Cid
+	selector ipld.Node
+	ls       *ipld.LinkSystem
+	opts     []Option
+}
+
+func (tc *traversalCar) WriteTo(w io.Writer) (int64, error) {
+	n, err := tc.WriteV2Header(w)
+	if err != nil {
+		return n, err
+	}
+	v1s, idx, err := tc.WriteV1(w)
+	n += int64(v1s)
+
+	if err != nil {
+		return n, err
+	}
+
+	// index padding, then index
+	opts := ApplyOptions(tc.opts...)
+	if opts.IndexCodec != index.CarIndexNoIndex {
+		if opts.IndexPadding > 0 {
+			buf := make([]byte, opts.IndexPadding)
+			pn, err := w.Write(buf)
+			n += int64(pn)
+			if err != nil {
+				return n, err
+			}
+		}
+		in, err := index.WriteTo(idx, w)
+		n += int64(in)
+		if err != nil {
+			return n, err
+		}
+	}
+
+	return n, err
+}
+
+func (tc *traversalCar) WriteV2Header(w io.Writer) (int64, error) {
+	n, err := w.Write(Pragma)
+	if err != nil {
+		return int64(n), err
+	}
+
+	h := NewHeader(tc.size)
+	conf := ApplyOptions(tc.opts...)
+	if p := conf.DataPadding; p > 0 {
+		h = h.WithDataPadding(p)
+	}
+	if p := conf.IndexPadding; p > 0 {
+		h = h.WithIndexPadding(p)
+	}
+	// TODO: support calculation / inclusion of the index.
+	h.IndexOffset = 0
+	hn, err := h.WriteTo(w)
+	if err != nil {
+		return int64(n) + hn, err
+	}
+	hn += int64(n)
+
+	// We include the initial data padding after the carv2 header
+	if h.DataOffset > uint64(hn) {
+		buf := make([]byte, h.DataOffset-uint64(hn))
+		n, err = w.Write(buf)
```

Comment:
This could be less than optimum in terms of memory consumption if the padding is large.
Use buffered write instead?

Reply:
added todo.